### ISSUE TITLE
Gemma sandwichNorms: eager-vs-traced StableHLO export-fidelity divergence (~1.4x/block)

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#763]
## Summary
For a real Gemma3 decoder built with `gemmaNetwork()`, the **`sandwichNorms`** feature (Gemma2/3 post-attention + post-FFN RMSNorms) produces a **numerical divergence between SKaiNET's eager executor and its own traced StableHLO export**. The exported function is a near-clean **scalar multiple** of the eager result, compounding **per block** (~1.4× per block). `qkNorm` and `layerOutputScale` do **not** have this problem — they export faithfully.

Surfaced by the conformance harness: SKaiNET-developers/skainet-iree-conformance#18 (the `gemma-decoder` model). Trace → `prunedToOutputs(logits)` → StableHLO → `iree-compile` (stock IREE 3.11.0) → `iree-run-module`, compared to the SKaiNET-CPU eager forward on identical weights.

## Evidence (bisection)
2-block decoder, dim 64, 2 heads, vocab 48, all weights deterministic, oracle = SKaiNET-CPU eager. `1/slope` = IREE-output / eager-output ratio from a linear fit:

| Config | 1/slope | R² | max_abs_err |
|---|--:|--:|--:|
| base block (qk/sandwich/output-scale **all off**) | 1.0018 | 1.00000 | **0.0023** ✅ |
| **qkNorm only** | 1.0018 | 1.00000 | 0.0023 ✅ |
| **layerOutputScale only** | 1.0021 | 1.00000 | 0.0024 ✅ |
| **sandwichNorms only** | **1.3976** | 0.99982 | **5.33** ❌ |

Block-count scaling with sandwichNorms on confirms per-block compounding: **1 block → 1.15×, 2 blocks → 1.56×**.

## Interpretation
The base decoder (embedding, RoPE causal attention, GeGLU, the *pre*-attention/*pre*-FFN RMSNorms, residuals) exports bit-close (≈2e-3), so the pre-norm RMSNorm path is faithful. Only the **sandwich (post) norms** diverge. The near-perfect linear/scalar signature points to the **RMSNorm gain** applied by the sandwich norms differing between the eager path and what the trace records / the converter lowers.

Worth checking:
- `RMSNormalization` `unitOffset` (gain = `1 + weight`) handling on the **traced** path for the sandwich-norm instances (the eager non-fused path at `RMSNormalization.kt` is taken for `unitOffset`, but the gain captured during trace / baked under `embedConstants=true` may drop the `+1`).
- `NeuralNetOperationsConverter.convertRmsNorm` applies plain `normalized * weight` with no unit-offset (though for the decoder path RMSNorm decomposes to primitives rather than a single `rmsNorm` node, so the decomposition is the more likely site).

## Repro
`skainet-iree-conformance` PR #18, model `gemma-decoder`: set `sandwichNorms = true` in `GemmaDecoderModel.kt` and run `USE_SKAINET_SOURCE=1 ./scripts/run-conformance.sh` — validate fails at max_abs_err ≈ 5.3; with `sandwichNorms = false` it passes at ≈2.4e-3.

**Comment by michalharakal** [comment]
### Deeper diagnosis (narrows the cause; disproves the obvious hypothesis)

Investigated via the `gemma-decoder` conformance model. Findings:

**1. The eager modules are correct and fully traceable — not the bug.**
- `RMSNormalization.forward` uses `ctx.ops` end-to-end, including the `unitOffset` gain (`ctx.ops.addScalar(w, 1f)`), so it traces faithfully.
- All Gemma norms are built with `unitOffset = false` (GemmaNetworkDef: `attn_norm`, `post_attention_norm`, `ffn_norm`, `post_ffw_norm`, `output_norm`) — so gain = `weight` directly; the `(1+w)` convention is **not** involved.
- `ResidualAdd.onForward` is `ctx.ops.add(input, skip)`; `HybridTransformerBlock.directForward`'s `residualBlockStarts` wires the sandwich skip connections correctly (verified by hand for the `[norm, mha, post_norm, residual, norm, ffn, post_norm, residual]` layout — no off-by-one). Eager and trace run the **same** `directForward`.

**2. RMSNorm lowering is exonerated, even for high-mean inputs.** Minimal repro: trace a single `RMSNormalization` (weight=1, eps=1e-6) over `[4,64]`, export → `iree-compile` (3.11.0) → run, vs eager:
- zero-mean input: `max_abs_err = 0.0000`
- **high-mean input (≈16)**: `max_abs_err = 0.0000`

So the earlier "variance vs mean-of-squares" hypothesis is **wrong** — the decomposed RMSNorm (`reduce`/`sqrt`/`divide`/`multiply`, no `custom_call`) is bit-exact regardless of input mean.

**3. The divergence only appears in the block composition WITH sandwich norms.**
- `gemma-decoder` with `sandwichNorms=false` (qk-norm + layer-output-scale on): validates at `max_abs_err ≈ 2.4e-3`.
- `sandwichNorms=true`: diverges. Bisected: it's `sandwichNorms` specifically (`qkNorm` and `layerOutputScale` are each faithful). Signature: a near-clean **per-block scalar ~1.4×** (1 block → 1.15×, 2 blocks → 1.56×), `R² ≈ 0.9998`, seq-independent. No `custom_call`/`reduce_mean` in the export.

Eager per-module dump (1 block, sandwich on) is sane: `post_attention_norm` correctly normalizes the high-magnitude attn output (mean≈15.7, rms≈16) to rms≈1.0; `post_ffw_norm` tames the FFN output (rms≈7900) to rms≈1.0.

**Conclusion / next step.** The components are individually correct; the bug is in the **trace → ComputeGraph → StableHLO** representation of the block *when a post-sublayer norm sits between the sublayer and its residual add* (the skip edge reaches back over an extra op). The decisive next probe is to prune the export to the first residual output (`afterAttn`) and compare eager-vs-IREE to localize attention-side vs FFN-side. (Workaround in `gemma-decoder`: omit `sandwichNorms`.)

**Comment by michalharakal** [comment]
### Further isolation (narrows trigger to constant-baking, but no minimal repro yet)

More probes via minimal hand-built repros (RMSNorm + matmul "sublayer" + residual), exported and run on IREE 3.11.0:

- **Generic sandwich, `embedConstants=false`** (`out = x + rmsnorm(matmul(rmsnorm(x), W))`, all tensors as args): **exact** (`max_abs_err = 0.0`), base and sandwich alike. So the skip-edge-over-an-intervening-norm is fine on the weights-as-args path.
- The real `gemma-decoder` uses **`embedConstants=true`** (RoPE cos/sin, masks, and RMSNorm weights bake as `stablehlo.constant`; only the unloaded weights stay args). The base block exports faithfully (`~2e-3`) but the **sandwich** block diverges (`~1.4×`/block).
- A minimal `embedConstants=true` repro with void placeholders for `x`/`W` and baked norm constants did **not** cleanly reproduce the sandwich-specific 1.4× (it showed a different, non-scale error in both base and sandwich, i.e. the toy setup doesn't faithfully mirror the real model's base — which validates).

**Net:** the failure needs the real model's combination — `embedConstants=true` constant-baking **+** post-(sublayer) norms **+** the real attention/FFN sublayers — together. RMSNorm itself, ResidualAdd, `directForward` wiring, and the generic sandwich structure are each individually correct. The most promising place to look is the **constant-baking / external-input synthesis in `toComputeGraph` (embedConstants=true)** when a baked-constant-weighted norm sits between a sublayer and its residual add. Decisive next probe (not yet done): prune the real export to the first residual output (`afterAttn`) and compare eager-vs-IREE to split attention-side vs FFN-side.

Workaround remains: omit `sandwichNorms` (done in `gemma-decoder`).

**Comment by michalharakal** [comment]
### Root-caused and fixed ✅

**Not a norm bug — a KV-cache trace-fidelity bug.** `PositionalKVCache.update` (transformer-core) copies the incoming K/V tensor **data** into a host `FloatArray` (`writeAt`) and reads it back via `ctx.fromData` (`sliceView`), **bypassing `ctx.ops`**. Under tracing the symbolic K/V carry no data, so `toComputeGraph(embedConstants=true)` baked an all-zero KV buffer as `stablehlo.constant` and disconnected the computed `k_proj`/`v_proj` — the exported decoder attended over **K=V=0**. Read straight off the MLIR:

```
%v4 = stablehlo.constant dense<0.0> : tensor<1x4x32xf32>   // K (baked-zero cache)
%v100 = SDPA(..., V=%v5 zeros) = 0  →  %v111 = o_proj(0) = 0  →  post_attn_norm(0)=0
%v125 = add(0, x) = x                                        // attn contributed nothing
```

Eager is fine (the buffer holds real data). In export it was **masked** in plain decoders — the unnormalized FFN (rms≈63000) dominates the residual stream, so zero-attention is negligible — and **exposed** by `sandwichNorms`: `post_ffw_norm` normalizes the FFN to rms≈1, so the lost attention becomes significant (~1.4×/block logit error). That's why only sandwich configs diverged.

(For the record, the earlier hypotheses — RMSNorm lowering, `(1+w)` unit-offset, the generic sandwich structure — were all ruled out; RMSNorm is bit-exact standalone even on high-mean inputs.)

**Fix:** `SKaiNET-developers/SKaiNET-transformers#193` — when `ctx.isRecording`, wire K/V functionally through `ops.concat` (the path `AppendKVCache` already uses) instead of the heap buffer; eager fast-path unchanged. The traced cache is now functional, so the graph also no longer surfaces dangling KV-cache leaves as outputs.

**Verified:** `skainet-iree-conformance` `gemma-decoder` — full Gemma3 block (qk-norm + RoPE + **sandwichNorms** + layer-output-scale): `export → iree-compile → iree-run-module` matches the SKaiNET-CPU oracle at **max_abs_err 3.8e-6** (was ~5.3). Harness PR: `SKaiNET-developers/skainet-iree-conformance#20`. Regression test `KvCacheTraceFidelityTest` added.

Note: the shared-cache variants (`SharedPositionalKVCache`, `PaddedSharedPositionalKVCache`, `OwnerReadOnlyKVCache`) still read via the buffer `sliceView` and need the same treatment for kv-shared model exports — follow-up.

Closing once #193 merges.
